src/cosmic_ray/execution/local.py

```diff
@@ -97,21 +101,23 @@ class LocalExecutionEngine(ExecutionEngine):
     "The local-git execution engine."
 
     def __call__(self, pending_work, config, on_task_complete):
-        with multiprocessing.Pool(
+        asyncio.run(self._execute_pending_works(pending_work, config, on_task_complete))
+
+    async def _execute_pending_works(self,
+                                     pending_work: Iterable[WorkItem],
+                                     config: ConfigDict,
+                                     on_task_complete: Callable):
+        loop = asyncio.get_running_loop()
+
+        with concurrent.futures.ProcessPoolExecutor(
                 initializer=_initialize_worker,
-                initargs=(config,)) as pool:
-
-            # pylint: disable=W0511
-            # TODO: This is not optimal. The pending-work iterable could be millions
-            # or billions of elements. We don't want to copy it. We copy it right
-            # now so that we don't access the database in a separate thread (i.e.
-            # one created by imap_unoredered below). We need to find a way around
-            # this.
-            pending = list(pending_work)
-
-            results = pool.imap_unordered(
-                func=_execute_work_item,
-                iterable=pending)
-
-            for job_id, result in results:
-                on_task_complete(job_id, result)
+                initargs=(config,)
+        ) as pool:
+
+            async def run_task(work_item):
+                result = await loop.run_in_executor(pool, _execute_work_item,
+                                                    work_item)
+                on_task_complete(*result)
+
+            tasks = [run_task(work_item) for work_item in pending_work]
+            await asyncio.gather(*tasks)
```

Comment:
I'm a bit concerned about memory usage here. `tasks` is a list of the same size as the number of items in `pending_work`, potentially millions or even billions of items. When we destructure it on the call to `asyncio.gather()`, Python is going to create a complete copy of it into a tuple, essentially doubling the space we need for this function.

Is there some way we can tell asyncio to process everything lazily and report when they're all done?